File: find-the-longest-substring-containing-vowels-in-even-counts.rs
// Find the Longest Substring Containing Vowels in Even Counts
impl Solution {
    pub fn find_the_longest_substring(s: String) -> i32 {
        let mut pos = vec![-1; 1<<5];
        pos[0] = 0;
        let mut sum = 0;
        let mut ans = 0;
        for (i, c) in s.as_bytes().iter().enumerate() {
            let i = i as i32;
            match c {
                b'a' => sum ^= 1,
                b'e' => sum ^= 2,
                b'i' => sum ^= 4,
                b'o' => sum ^= 8,
                b'u' => sum ^= 16,
                _ => (),
            }
            if pos[sum] < 0 {
                pos[sum] = i+1;
            } else {
                ans = ans.max(i+1-pos[sum]);
            }
        }
        ans
    }
}

File: find-the-longest-substring-containing-vowels-in-even-counts.nim
# Find the Longest Substring Containing Vowels in Even Counts
import std/algorithm

proc findTheLongestSubstring(s: cstring): int {.exportc.} =
  var
    bgn: array[32, int]
    m = 0
    i = 0
  bgn.fill(-1)
  bgn[0] = 0
  while s[i] != '\0':
    case s[i]
    of 'a': m = m xor 1
    of 'e': m = m xor 2
    of 'i': m = m xor 4
    of 'o': m = m xor 8
    of 'u': m = m xor 16
    else: discard
    inc i
    if bgn[m] >= 0:
      result = max(result, i-bgn[m])
    else:
      bgn[m] = i
